src/storage/src/compaction/task.rs

```diff
@@ -155,19 +173,51 @@ impl<S: LogStore> CompactionTask for CompactionTaskImpl<S> {
             "Compacting SST files, input: {:?}, output: {:?}, window: {:?}",
             input_ids, output_ids, self.compaction_time_window
         );
-        self.write_manifest_and_apply(output, compacted)
+
+        let no_output = output.is_empty();
+        let write_result = self
+            .write_manifest_and_apply(output, compacted)
             .await
             .map_err(|e| {
                 error!(e; "Failed to update region manifest: {}", self.shared_data.name());
                 e
-            })
+            });
+
+        if !no_output && self.reschedule_on_finish {
+            // only reschedule another compaction if current compaction has output and it's
+            // triggered by flush.
+            if let Err(e) = self
+                .writer
+                .compact(WriterCompactRequest {
+                    shared_data: self.shared_data.clone(),
+                    sst_layer: self.sst_layer.clone(),
+                    manifest: self.manifest.clone(),
+                    wal: self.wal.clone(),
+                    region_writer: self.writer.clone(),
+                    compact_ctx: CompactContext {
+                        wait: false,
+                        max_files_in_l0: 8,
+                    },
```

Comment:
Could we avoid hard-coding max_files_in_l0? Maybe we could make compact_ctx optional or max_files_in_l0 optional.